adapters/interlocked_win32.c

```diff
@@ -1,4 +1,330 @@
 // Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
+#include <stdint.h>
+#include "windows.h"
+#include "azure_c_util/interlocked.h"
 
-#include "azure_c_util/interlocked.h"
\ No newline at end of file
+IMPLEMENT_MOCKABLE_FUNCTION(, int32_t, interlocked_add, volatile int32_t*, addend, int32_t, value) 
+{
+    /*Codes_SRS_INTERLOCKED_43_001 [ interlocked_add shall atomically add *addend with value and store the result in *addend.]*/
+    /*Codes_SRS_INTERLOCKED_43_032: [ interlocked_add shall return the result of the addition.]*/
+    /*Codes_SRS_INTERLOCKED_WIN32_43_001: [ interlocked_add shall call InterlockedAdd from windows.h. ]*/
+    /*Codes_SRS_INTERLOCKED_WIN32_43_002: [ interlocked_add shall return the result of the addition. ]*/
+
+    return InterlockedAdd((volatile LONG*)addend, value);
```

Comment:
>(volatile LONG*) [](start = 26, length = 16)

why is the cast necessary?

Reply:
Here and everywhere. Point being: either the cast "does something" (changes value / signedness etc) (in which case it is wrong), or it doesn't do anything, in which case it is extraneous. 

There are legitimate cases where casts are necessary. Is this necessary and if so... why... I asked this in iteration 6 too - looking back at the comments,

---
In reply to: [430678730](https://github.com/Azure/azure-c-util/pull/43#discussion_r430678730) [](ancestors = 430678730)

Reply:
int32_t is typedefed to int (on win32 at least). LONG is long.
There is definitely a warning assigning between long and int even though they are 32-bit signed integers on Windows. I think that's why these casts remain.

---
In reply to: [430679721](https://github.com/Azure/azure-c-util/pull/43#discussion_r430679721) [](ancestors = 430679721,430678730)